server/src/main/java/org/opensearch/gateway/remote/RemoteClusterStateCleanupManager.java

```diff
@@ -0,0 +1,365 @@
+        "cluster.remote_store.state.cleanup_interval",
+        CLUSTER_STATE_CLEANUP_INTERVAL_DEFAULT,
+        CLUSTER_STATE_CLEANUP_INTERVAL_MINIMUM,
+        Setting.Property.NodeScope,
+        Setting.Property.Dynamic
+    );
+    private static final Logger logger = LogManager.getLogger(RemoteClusterStateCleanupManager.class);
+    private RemoteClusterStateService remoteClusterStateService;
+    private RemotePersistenceStats remoteStateStats;
+    private BlobStoreTransferService blobStoreTransferService;
+    private volatile TimeValue staleFileCleanupInterval;
+    private final AtomicBoolean deleteStaleMetadataRunning = new AtomicBoolean(false);
+    private AsyncStaleFileDeletion staleFileDeletionTask;
+    private long lastCleanupAttemptStateVersion;
+    private final ThreadPool threadpool;
+    private final ClusterApplierService clusterApplierService;
+
+    public RemoteClusterStateCleanupManager(RemoteClusterStateService remoteClusterStateService, ClusterService clusterService) {
+        this.remoteClusterStateService = remoteClusterStateService;
+        this.remoteStateStats = remoteClusterStateService.getStats();
+        ClusterSettings clusterSettings = clusterService.getClusterSettings();
+        this.clusterApplierService = clusterService.getClusterApplierService();
+        this.staleFileCleanupInterval = clusterSettings.get(REMOTE_CLUSTER_STATE_CLEANUP_INTERVAL_SETTING);
+        this.threadpool = remoteClusterStateService.getThreadpool();
+        // initialize with 0, a cleanup will be done when this node is elected master node and version is incremented more than threshold
+        this.lastCleanupAttemptStateVersion = 0;
+        clusterSettings.addSettingsUpdateConsumer(REMOTE_CLUSTER_STATE_CLEANUP_INTERVAL_SETTING, this::updateCleanupInterval);
+    }
+
+    void start() {
+        staleFileDeletionTask = new AsyncStaleFileDeletion(this);
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (staleFileDeletionTask != null) {
+            staleFileDeletionTask.close();
+        }
+    }
+
+    private BlobStoreTransferService getBlobStoreTransferService() {
+        if (blobStoreTransferService == null) {
+            blobStoreTransferService = new BlobStoreTransferService(remoteClusterStateService.getBlobStore(), threadpool);
+        }
+        return blobStoreTransferService;
+    }
+
+    private void updateCleanupInterval(TimeValue updatedInterval) {
+        this.staleFileCleanupInterval = updatedInterval;
+        logger.info("updated remote state cleanup interval to {}", updatedInterval);
+        // After updating the interval, we need to close the current task and create a new one which will run with updated interval
+        if (staleFileDeletionTask != null && !staleFileDeletionTask.getInterval().equals(updatedInterval)) {
+            staleFileDeletionTask.setInterval(updatedInterval);
+        }
+    }
+
+    void cleanUpStaleFiles() {
```

Comment:
why package private ?

Reply:
This is visible for testing, using this is unit tests